File: bin/correct_logR_ichorcna.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(argparser, quietly = TRUE)
  library(readr, quietly = TRUE)
  library(dplyr, quietly = TRUE)
}
)


adjust_log_ratio <- function(copy_number, ploidy, min.ratio = 2^-8) {

    adjusted <- copy_number / ploidy
    adjusted <- pmax(min.ratio, adjusted)
    adjusted <- log2(adjusted)
    return(adjusted)
}

parser <- arg_parser("Adjust logR.", hide.opts = TRUE)

parser <- add_argument(parser, "--seg",
                       help = "Segmentation file from which extract information.", #nolint
                       nargs = Inf)

parser <- add_argument(parser, "--ploidy",
                       help = "Ploidy Summary file from which extract information.", #nolint
                       nargs = Inf)

parser <- add_argument(parser, "--project",
                       help = "Samplename")

args <- parse_args(parser)

message("Adjusting log-Ratio...")


# TO DO: Conditional correction of logR

# IMPORT TWO DATAFRAMES
df <- read_tsv(args$seg)
df_ploidy <- read_tsv(args$ploidy)

#RENAME COLUMNS

df <- df %>%  rename(all_of(c(chromosome = "chrom", sample = "ID")))

df_ploidy <- df_ploidy %>%
  rename(all_of(c(purity = "Tumor Fraction",
                  ploidy = "Ploidy",
                  sample = "samplename")))

# CREATE A TEMPORARY COMPLETE DF TO CORRECT LOGR
df_tmp <- df %>% select(sample) %>% left_join(df_ploidy, by = 'sample') %>% select(-sample)
df <- cbind(df, df_tmp)

# adjusted logR (https://github.com/lima1/PureCN/issues/40)
#Solution in:
#https://github.com/maitnnguyen/Oseq_CNV_Paper/blob/main/src/pipeline/step4_CNVcall.R # nolint

df <- df %>%
    dplyr::mutate(adj.seg = adjust_log_ratio(logR_Copy_Number,
        ploidy = df_ploidy$ploidy))

# df$adj.seg <- (df$purity*df$ploidy*R + 2*(1-df$purity)*(R - 1))/(df$purity*df$ploidy) #nolint

# FINAL DF WITH SELECTED COLUMNS
df_gistic <- df %>%
            dplyr::select(sample, chromosome, start, end, num.mark, adj.seg) %>%
            rowwise %>%
            filter(!is.infinite(adj.seg))

readr::write_tsv(df_gistic,
                file = "segments_logR_corrected_gistic.seg",
                quote = "needed")

File: bin/correct_logr_ichorcna.py
#!/usr/bin/env python

from pathlib import Path

import polars as pl
import typer

from typing import Optional
from typing_extensions import Annotated

__version__ = "0.1.0"


def version_callback(value: bool):
    if value:
        print(f"correct_logr_ichorcna {__version__}")
        raise typer.Exit()


def main(
    seg: Path = typer.Option(
        default=...,
        help="Segmentation file from which extract information",
    ),
    ploidy: Path = typer.Option(
        default=...,
        help="Ploidy summary file from which extract information",
    ),
    version: bool | None = typer.Option(
        default=None, callback=version_callback
    ),
):

    segments = pl.scan_csv(seg, separator="\t").rename({"ID": "sample"})
    ploidy_df = pl.scan_csv(ploidy, separator="\t")

    result = (
        segments.join(
            ploidy_df.select(pl.col("samplename").alias("sample"), "Ploidy"),
            on="sample",
            how="left",
        )
        .with_columns(
            pl.col("logR_Copy_Number")
            .truediv(pl.col("Ploidy"))
            .clip(lower_bound=2e-8)
            .log(base=2)
            .clip(lower_bound=-0.5)
            .alias("adj.seg")
        )
        .select("sample", "chromosome", "start", "end", "num.mark", "adj.seg")
    )

    result.collect().write_csv(
        "segments_logR_corrected_gistic.seg",
        separator="\t",
        quote_style="necessary",
    )


if __name__ == "__main__":
    typer.run(main)
